cc/utils/cose/cose.h

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2024 The Project Oak Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CC_UTILS_COSE_COSE_H_
+#define CC_UTILS_COSE_COSE_H_
+
+#include <memory>
+#include <string>
+
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+#include "libcppbor/include/cppbor/cppbor.h"
+#include "libcppbor/include/cppbor/cppbor_parse.h"
+
+namespace oak::utils::cose {
+
+// COSE_Sign1 object.
+// <https://datatracker.ietf.org/doc/html/rfc8152#section-4.2>
+struct CoseSign1 {
+  // Parameters about the current layer that are to be cryptographically protected.
+  const cppbor::Bstr* protected_headers;
+  // Parameters about the current layer that are not cryptographically protected.
+  const cppbor::Map* unprotected_headers;
+  // Serialized content to be signed.
+  const cppbor::Bstr* payload;
+  // Array of signatures. Each signature is represented as a COSE_Signature structure.
+  const cppbor::Bstr* signature;
+
+  CoseSign1(const cppbor::Bstr* protected_headers, const cppbor::Map* unprotected_headers,
+            const cppbor::Bstr* payload, const cppbor::Bstr* signature,
+            std::unique_ptr<cppbor::Item>&& item)
+      : protected_headers(protected_headers),
+        unprotected_headers(unprotected_headers),
+        payload(payload),
+        signature(signature),
+        item_(std::move(item)) {}
+
+  static absl::StatusOr<CoseSign1> Deserialize(const std::vector<uint8_t>& data);
+
+ private:
+  // Parsed CBOR item containing COSE_Sign1 object.
+  std::unique_ptr<cppbor::Item> item_;
+};
+
+// COSE_Key object.
+// <https://www.rfc-editor.org/rfc/rfc8152#section-7>
+struct CoseKey {
+  // Identification of the key type.
+  const cppbor::Uint* kty;
+  // Key identification value.
+  const cppbor::Bstr* kid;
+  // Key usage restriction to this algorithm.
+  const cppbor::Nint* alg;
+  // Restrict set of permissible operations.
+  const cppbor::Array* key_ops;
+
+  // EC identifier.
+  const cppbor::Uint* crv;
+  // Public key.
+  const cppbor::Bstr* x;
+
+  CoseKey(const cppbor::Uint* kty, const cppbor::Bstr* kid, const cppbor::Nint* alg,
```

Comment:
Thinking API-wise, and this is in line with my comments above...

I'd change these constructors to just take `cppbor::Item` directly. Once you've assigned it to item_, then, in the constructor you can build the pointers to inside the item.

Or: move the constructors to the private section so that the only way how to construct these structs is via the factory method, which enforces the invariants.